google-ads-examples/src/main/java/com/google/ads/googleads/examples/misc/NavigateSearchResultPagesCachingTokens.java

```diff
@@ -0,0 +1,247 @@
+ * at a time without caching all the results first.
+ *
+ * <p>To add campaigns, {@link com.google.ads.googleads.examples.basicoperations.AddCampaigns}.
+ */
+public class NavigateSearchResultPagesCachingTokens {
+
+  // The maximum number of results to retrieve.
+  private static final int RESULTS_LIMIT = 10;
+  // The size of the paginated search result pages.
+  private static final int PAGE_SIZE = 3;
+
+  private static class NavigateSearchResultPagesCachingTokensParams extends CodeSampleParams {
+
+    @Parameter(names = ArgumentNames.CUSTOMER_ID, required = true)
+    private Long customerId;
+  }
+
+  public static void main(String args[]) {
+    NavigateSearchResultPagesCachingTokensParams params =
+        new NavigateSearchResultPagesCachingTokensParams();
+    if (!params.parseArguments(args)) {
+      // Either pass the required parameters for this example on the command line, or insert them
+      // into the code here. See the parameter class definition above for descriptions.
+      params.customerId = Long.parseLong("ENTER_CUSTOMER_ID_HERE");
+    }
+
+    GoogleAdsClient googleAdsClient = null;
+    try {
+      googleAdsClient = GoogleAdsClient.newBuilder().fromPropertiesFile().build();
+    } catch (FileNotFoundException fnfe) {
+      System.err.printf(
+          "Failed to load GoogleAdsClient configuration from file. Exception: %s%n", fnfe);
+      System.exit(1);
+    } catch (IOException ioe) {
+      System.err.printf("Failed to create GoogleAdsClient. Exception: %s%n", ioe);
+      System.exit(1);
+    }
+
+    try {
+      new NavigateSearchResultPagesCachingTokens().runExample(googleAdsClient, params.customerId);
+    } catch (GoogleAdsException gae) {
+      // GoogleAdsException is the base class for most exceptions thrown by an API request.
+      // Instances of this exception have a message and a GoogleAdsFailure that contains a
+      // collection of GoogleAdsErrors that indicate the underlying causes of the
+      // GoogleAdsException.
+      System.err.printf(
+          "Request ID %s failed due to GoogleAdsException. Underlying errors:%n",
+          gae.getRequestId());
+      int i = 0;
+      for (GoogleAdsError googleAdsError : gae.getGoogleAdsFailure().getErrorsList()) {
+        System.err.printf("  Error %d: %s%n", i++, googleAdsError);
+      }
+      System.exit(1);
+    }
+  }
+
+  /**
+   * Runs this example.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @throws GoogleAdsException if an API request failed with one or more service errors.
+   */
+  private static void runExample(GoogleAdsClient googleAdsClient, long customerId) {
+    // The cache of page tokens. It is stored in-memory and in ascendant order of page number.
+    // The first page's token is always an empty string.
+    List<String> pageTokens = new ArrayList<>(Collections.singletonList(""));
```

Comment:
Did you consider using a `SortedMap<Integer, String>` for this? I think that would be more intuitive for a cache, and would avoid the mental exercise of translating 0-based list indices to 1-based page numbers. I think that'd make `cacheNextPageToken()` and lines 183-194 simpler.

You'd still be able to iterate over the keys (page numbers) in order with a `SortedMap`, and could also use `headMap` or `tailMap` if you need all keys `< someKey` or all keys `>= someKey`, respectively.

Reply:
I went back and forth on this using a map, but didn't think to using a SORTED map - thanks!